File: 572. Subtree of Another Tree.cpp
#include<vector>
#include<iostream>
#include<stack>

using namespace std;


struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};



class Solution {
    public:

        bool checkForSubtree(TreeNode* root, TreeNode* subRoot){
            
            if(!root && !subRoot){
                return true;
            }

            if(!root || !subRoot || root->val != subRoot->val){
                return false;
            }

            return checkForSubtree(root->left,subRoot->left) && checkForSubtree(root->right,subRoot->right);

        }


        bool isSubtree(TreeNode* root, TreeNode* subRoot) {
            // Perform dfs till value is same than call subtree
                
            stack<TreeNode*> s;

            s.push(root);

            while(s.size()>0){
                
                TreeNode* cur = s.top();

                s.pop();

                if(cur->val == subRoot->val){
                    if(checkForSubtree(cur,subRoot)){
                        return true;
                    }
                }

                if(cur->left){
                    s.push(cur->left);
                }

                if(cur->right){
                    s.push(cur->right);
                }

            }
            
            return false;
        }
};

int main(){

    Solution solution;

    TreeNode* root = new TreeNode(3);
    root->right = new TreeNode(5);
    

    TreeNode* main_2 = new TreeNode(4);
    main_2->left = new TreeNode(1);
    main_2->right = new TreeNode(2);

    TreeNode* subTreeRoot = new TreeNode(4);

    subTreeRoot->left = new TreeNode(1);
    subTreeRoot->right = new TreeNode(2);

    root->left = main_2;


     bool answer = solution.isSubtree(root,subTreeRoot);

    cout << "Answer: " << answer << endl;

}

File: 572. Subtree of Another Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        
        def dfs(root, subRoot) -> bool:
            if root == None and subRoot == None: return True
            elif root == None or subRoot == None: return False
            elif root.val != subRoot.val: return False
            return dfs(root.left,subRoot.left) and dfs(root.right,subRoot.right)

        s = collections.deque()
        s.append(root)
        while s:
            cur = s.pop()
            if cur == None: continue
            if cur.val == subRoot.val: 
                if dfs(cur,subRoot): return True
            s.append(cur.left)
            s.append(cur.right)
            
        return False